File: C/LinkedList/04_Doubly_Circular_LinkedList_C/DoublyCL_01.c
/*
===============================================================================
Program Name    : Doubly Circular Linked List - Node Structure and Memory Size
Program Number  : 001
Description     : This program introduces the basic node structure used in a
                  Doubly Circular Linked List. It demonstrates the creation
                  of a self-referential structure containing data, next, and
                  previous pointers, and displays the memory occupied by a
                  single node using the sizeof operator.

Date            : July 2026
Language        : C
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Structure
2. Self-Referential Structure
3. Pointer Declaration
4. typedef
5. sizeof Operator
6. Structure Packing using #pragma pack(1)
7. Doubly Circular Node Representation

Key Learning Points:
- A node contains data, next pointer, and previous pointer.
- next stores the address of the next node.
- prev stores the address of the previous node.
- #pragma pack(1) removes padding bytes.
- sizeof() returns the total memory occupied by a node.
- typedef improves code readability.

===============================================================================
*/

#include<stdio.h>

#pragma pack(1)
struct node
{
    int data;
    struct node *next;
    struct node *prev;
};
typedef struct node NODE;
typedef struct node *PNODE;
typedef struct node **PPNODE;

int main()
{
    NODE obj;

    printf("Size of node is : %d\n",(int)sizeof(obj));

    return 0;
}

/*
===============================================================================
Expected Output:
===============================================================================

Size of node is : 20

(Note: Output may vary depending on the compiler, operating system,
system architecture, and structure packing.)

===============================================================================
Memory Representation:
===============================================================================

Node Structure

+------------+----------------------+----------------------+
| data (4B)  | next pointer (8B)    | prev pointer (8B)    |
+------------+----------------------+----------------------+

Total Size (with #pragma pack(1)) = 20 Bytes

Without #pragma pack(1)

+------------+---------+----------------------+----------------------+
| data (4B)  | Padding | next pointer (8B)    | prev pointer (8B)    |
+------------+---------+----------------------+----------------------+

Total Size = 24 Bytes (Compiler Dependent)

===============================================================================
Pointer Explanation:
===============================================================================

struct node
{
    int data;
    struct node *next;
    struct node *prev;
};

data
└── Stores the integer value.

next
└── Stores the address of the next node.

prev
└── Stores the address of the previous node.

typedef Usage

NODE
└── Represents struct node

PNODE
└── Pointer to a node

PPNODE
└── Pointer to pointer to a node

===============================================================================
Algorithm:
===============================================================================

1. Define the node structure.
2. Declare typedefs for easy usage.
3. Create one node object.
4. Calculate memory occupied using sizeof().
5. Display the node size.
6. Stop.

===============================================================================
Program Flow:
===============================================================================

          Start
            │
            ▼
  Define Node Structure
            │
            ▼
   Create Node Object
            │
            ▼
 Calculate sizeof(node)
            │
            ▼
 Display Node Size
            │
            ▼
           Stop

===============================================================================
Time Complexity:
===============================================================================

sizeof()             : O(1)

Space Complexity     : O(1)

===============================================================================
Conclusion:
===============================================================================

Program 001 introduces the basic node structure required for implementing a
Doubly Circular Linked List in C. It demonstrates the use of a self-
referential structure containing data, next, and previous pointers, along
with typedef declarations and the sizeof operator to understand the memory
occupied by a node. This program forms the foundation for all subsequent
Doubly Circular Linked List operations.

===============================================================================
*/

File: Java/LinkedList/04_Doubly_Circular_LinkedList_Java/DoublyCL_01.java
/*
===============================================================================
Program Name    : Doubly Circular Linked List - Node Class & Object Creation
Program Number  : 001
File Name       : DoublyCL_01.java
Description     : This program demonstrates the creation of a node for a
                  Doubly Circular Linked List using Java.

Date            : July 2026
Language        : Java
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Class
2. Object-Oriented Programming (OOP)
3. Self-Referential Class
4. Constructor
5. Objects

Key Learning Points:
- Create a node for a Doubly Circular Linked List.
- Understand self-referential classes.
- Initialize data, next, and previous references.
- Create an object of the Node class.

===============================================================================
*/

class Node
{
    public int data;
    public Node next;
    public Node prev;

    public Node(int no)
    {
        data = no;
        next = null;
        prev = null;
    }
}

class DoublyCL_01
{
    public static void main(String A[])
    {
        Node obj = new Node(11);

        System.out.println("Data : " + obj.data);
        System.out.println("Next : " + obj.next);
        System.out.println("Previous : " + obj.prev);
    }
}
/*
===============================================================================
Expected Output:
===============================================================================

Data : 11
Next : null
Previous : null

===============================================================================
Memory Representation:
===============================================================================

        obj
         │
         ▼

+--------------------------------+
| Prev | Data | Next |
| null |  11  | null |
+--------------------------------+

===============================================================================
Algorithm:
===============================================================================

1. Define the Node class.
2. Declare data, next, and previous members.
3. Create a constructor to initialize the node.
4. Create an object of the Node class.
5. Display the data and reference values.
6. Stop.

===============================================================================
Program Flow:
===============================================================================

          Start
             │
             ▼
      Define Node Class
             │
             ▼
      Create Constructor
             │
             ▼
      Create Node Object
             │
             ▼
     Display Node Details
             │
             ▼
            Stop

===============================================================================
Time Complexity:
===============================================================================

Node Creation : O(1)

Space Complexity : O(1)

===============================================================================
Conclusion:
===============================================================================

Program 001 demonstrates the basic structure of a node in a Doubly Circular
Linked List using Java. Each node stores data along with references to the
next and previous nodes. This forms the foundation for implementing all
subsequent Doubly Circular Linked List operations.

===============================================================================
*/
